XAlloc.h

```diff
@@ -5,34 +5,88 @@
+#include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 
-void fail(void) ATTR_NORETURN;
+#define write_failure(len, ret, filename, lineno)       \
+   do {                                                 \
+      CRT_done();                                       \
+      fprintf(stderr,                                   \
+         "Failed to write memory (%zu/%zu) [%s:%d]\n",  \
+         (size_t)len,                                   \
+         (size_t)ret,                                   \
+         filename,                                      \
+         lineno);                                       \
+      abort();                                          \
+   } while(0)
 
-void* xMalloc(size_t size);
+#define oom_failure(filename, lineno)            \
+   do {                                          \
+      CRT_done();                                \
+      fprintf(stderr,                            \
+         "Failed to allocate memory [%s:%d]\n",  \
+         filename,                               \
+         lineno);                                \
+      abort();                                   \
+   } while(0)
 
-void* xCalloc(size_t nmemb, size_t size);
+static inline void* xMalloc_impl(size_t size, const char *filename, unsigned int lineno) {
+   void *ret = malloc(size);
+   if (!ret) {
+      oom_failure(filename, lineno);
+   }
+   return ret;
+}
+#define xMalloc(size) xMalloc_impl(size, __FILE__, __LINE__)
 
-void* xRealloc(void* ptr, size_t size);
 
-#undef xAsprintf
+static inline void* xCalloc_impl(size_t nmemb, size_t size, const char *filename, unsigned int lineno) {
+   void *ret = calloc(nmemb, size);
+   if (!ret) {
+      oom_failure(filename, lineno);
+   }
+   return ret;
+}
+#define xCalloc(nmemb, size) xCalloc_impl(nmemb, size, __FILE__, __LINE__)
 
-#define xAsprintf(strp, fmt, ...) do { int _r=asprintf(strp, fmt, __VA_ARGS__); if (_r < 0) { fail(); } } while(0)
+static inline void* xRealloc_impl(void *ptr, size_t size, const char *filename, unsigned int lineno) {
+   void *ret = realloc(ptr, size);
+   if (!ret) {
+      oom_failure(filename, lineno);
+   }
+   return ret;
+}
+#define xRealloc(ptr, size) xRealloc_impl(ptr, size, __FILE__, __LINE__)
 
-#define xSnprintf(fmt, len, ...) do { int _l=len; int _n=snprintf(fmt, _l, __VA_ARGS__); if (!(_n > -1 && _n < _l)) { curs_set(1); endwin(); err(1, NULL); } } while(0)
+static inline char* xStrdup(const char *str,  const char *filename, unsigned int lineno) {
+   assert(str);
+   char *ret = strdup(str);
+   if (!ret) {
+      oom_failure(filename, lineno);
+   }
+   return ret;
+}
+#define xStrdup(str) xStrdup(str, __FILE__, __LINE__)
 
-#undef xStrdup
-#undef xStrdup_
-#ifdef NDEBUG
-# define xStrdup_ xStrdup
-#else
-# define xStrdup(str_) (assert(str_), xStrdup_(str_))
-#endif
+#define xAsprintf(strp, fmt, ...)                   \
+   do {                                             \
+      int _ret = asprintf(strp, fmt, __VA_ARGS__);  \
```

Comment:
Prefer `vasprintf` as shown with my implementation in #155.

Reply:
There is no need inside a macro for manual argument unpacking